docs: add guide for displaying dates without hydration flash (#92786)
vercel/next.js: docs/01-app/02-guides/preventing-flash-before-hydration.mdx

---
title: How to prevent flash before hydration
nav_title: Preventing Flash
description: Learn how to correct server-rendered content before the browser paints, avoiding visible flash when the page hydrates.
---

> Some values can only be formatted correctly in the browser. A date needs the user's locale and timezone. A theme depends on a preference saved in `localStorage`. The server doesn't have access to these, so the HTML it sends is a best guess.

Reviewer: Getting a bit AI language vibe here on these sentences

Reviewer 2: yeah the initial PR had some stuff... I did a pass already though so this is 60% me lol

Using an **inline script** that runs synchronously as the browser parses the HTML, you can correct the DOM **before the first paint**. This guide covers dates, themes, and persisted UI state.

## Example

The companion [demo](https://preventing-flash-before-hydration.labs.vercel.dev/) ([source](https://github.com/vercel-labs/preventing-flash-before-hydration)) is built with `LANG=ja_JP.UTF-8` to simulate a server/client locale mismatch. It lets you compare:

- Inline script date formatting with no flash and no hydration error
- A Client Component that calls `toLocaleDateString()` directly, showing the hydration error
- An accordion that persists its open section to `localStorage` using a lazy `useState` initializer

Use Chrome DevTools [Sensors](https://developer.chrome.com/docs/devtools/sensors) to override your locale (e.g. `ru-RU`) and see how each approach behaves.

## Dates and formatting

> A timestamp like `2026-06-15T18:00:00Z` is the same instant everywhere. But to be useful to the reader, you'll want to format it for their locale and timezone. When you call `toLocaleDateString()` or `Intl.DateTimeFormat` on the server, they format using the **server's** settings, not the user's.

Reviewer: "is the same instant everywhere"

what does this mean?

### The problem

A natural approach is a Client Component that formats the date:

```tsx filename="app/components/event-date.tsx" switcher
'use client'

export function EventDate({ date }: { date: string }) {
  return <p>{new Date(date).toLocaleDateString()}</p>
}
```

```jsx filename="app/components/event-date.js" switcher
'use client'

export function EventDate({ date }) {
  return <p>{new Date(date).toLocaleDateString()}</p>
}
```

During SSR, `toLocaleDateString()` runs in Node.js and formats using the server's locale (e.g. `6/15/2026`). On hydration, React re-executes the component in the browser and produces the user's locale (e.g. `2026/6/15`). React detects the mismatch, throws a hydration error, and the user sees a flash.

You could work around this with `useEffect` to update the date after hydration, but the user still sees the server-formatted date first. A Server Component avoids the error entirely, but the date is permanently in the server's locale.

> > **Good to know:** This issue is easy to miss in local development if your machine's locale matches your browser's. If you typically deploy to a different region, run your dev server with `TZ` and `LANG` set to a different locale than your browser (e.g. `LANG=ja_JP.UTF-8 next dev` or `LANG=en_US.UTF-8 next dev`) to catch these mismatches early.

Reviewer 3: This is good, I'd also add `TZ=UTC` as an example. I always do that because most servers are UTC (and so it catches hydration errors early).

import { getEvent } from '@/app/lib/events'

export default async function Page() {
  const event = await getEvent('nextjs-conf')

  return (
    <section>
      <h1>{event.name}</h1>
      <p id="event-date" suppressHydrationWarning>
        {new Date(event.date).toLocaleDateString()}
      </p>
      <script
        dangerouslySetInnerHTML={{
          __html: `document.getElementById("event-date").textContent=new Date("${event.date}").toLocaleDateString()`,
        }}
      />
    </section>
  )
}
```

```jsx filename="app/events/page.js" switcher
import { getEvent } from '@/app/lib/events'

export default async function Page() {
  const event = await getEvent('nextjs-conf')

  return (
    <section>
      <h1>{event.name}</h1>
      <p id="event-date" suppressHydrationWarning>
        {new Date(event.date).toLocaleDateString()}
      </p>
      <script
        dangerouslySetInnerHTML={{
          __html: `document.getElementById("event-date").textContent=new Date("${event.date}").toLocaleDateString()`,
        }}
      />
    </section>
  )
}
```

1. The server renders the date with its locale into the `<p>`. No component code is sent to the browser.

> 2. The inline script runs **synchronously** during HTML parsing and replaces the text.

Reviewer: this explanation seems duplicated in the segment above the snippet, maybe just keep it here and remove it above?

    <html lang="en" data-theme="light" suppressHydrationWarning>
      <head>
        <script
          dangerouslySetInnerHTML={{
            __html: `(function(){try{var t=localStorage.getItem("theme");if(t)document.documentElement.setAttribute("data-theme",t)}catch(e){}})()`,
          }}
        />
      </head>
      <body>{children}</body>
    </html>
  )
}
```

```jsx filename="app/layout.js" switcher
export default function RootLayout({ children }) {
  return (
    <html lang="en" data-theme="light" suppressHydrationWarning>
      <head>
        <script
          dangerouslySetInnerHTML={{
            __html: `(function(){try{var t=localStorage.getItem("theme");if(t)document.documentElement.setAttribute("data-theme",t)}catch(e){}})()`,
          }}
        />
      </head>
      <body>{children}</body>
    </html>
  )
}
```

```css filename="app/globals.css"
[data-theme='light'] {
  --background: #ffffff;
  --foreground: #000000;
}

[data-theme='dark'] {
  --background: #0a0a0a;
  --foreground: #ededed;
}
```

The script runs in `<head>`, so the correct theme is applied before any content is painted. The `try/catch` handles cases where `localStorage` is unavailable.

> > **Good to know:** If the theme preference is stored in a cookie, read it on the server in a [layout](/docs/app/api-reference/file-conventions/layout) or [middleware](/docs/app/building-your-application/routing/middleware) and render the correct theme directly.

Reviewer: Hmm, thing brings up issues with CC. The type of thing [precompute](https://aurorascharff.no/posts/the-precompute-pattern-encoding-dynamic-data-into-urls-in-nextjs/) or https://next-recipes-git-static-variants-97f72d-sam-selikoffs-projects.vercel.app/static-page-variants would solve right? Do we have any reccomendations?

Reviewer 2: Oh I do have one - in this case, you just do the same as localStorage, read the cookie client side in that inline script

Reviewer: Yep I think that's what Ive seen recommended before! But it might be good to document too. We def should have a guide on these.  
  
\- static variants  
\-precomute  
\-inline scripts  
  
things to improve CC?

---
title: How to prevent flash before hydration
nav_title: Preventing Flash
description: Learn how to correct server-rendered content before the browser paints, avoiding visible flash when the page hydrates.
---

User preferences, browser settings, and client-side storage aren't available during server rendering. The server emits a reasonable default, but any UI that depends on client-only state (locale, timezone, theme, persisted interactions) needs to be updated before the user sees it.

> Common approaches to this problem each have tradeoffs. A Client Component that re-renders with client values causes a hydration error. Deferring to `useEffect` avoids the error but introduces a visible flash. And rendering only on the server means giving up on client-specific formatting entirely.

Reviewer: maybe we should use the common docs pattern of the bullets?

common approaches:

- 
- 
- 

bla bla tradeoffs?

Reviewer 2: Mmm actually, we can imply the trade offs into each bulletpoint

```suggestion
Some common approaches to this problem:

- A Client Component that re-renders with client values causes a hydration error.
- Deferring to `useEffect` avoids the error but introduces a visible flash.
- Rendering only on the server means giving up on client-specific formatting entirely.
```
